File: apps/cli/surface_client.py
"""Stdlib loopback Surface client for the local Agent OS Runtime daemon.

Every state-changing call sends a versioned command with a fresh idempotency
key and the tracked event sequence for that session; every success is parsed
through the matching closed Pydantic contract. HTTP and network failures map
to closed client exceptions that never expose the bearer token or raw headers.
"""

from __future__ import annotations

import json
import socket
import urllib.error
import urllib.request
from datetime import datetime, timezone
from typing import Literal
from uuid import uuid4

from agent_os_contracts import (
    SURFACE_PROTOCOL_VERSION,
    ApprovalDisposition,
    ContractModel,
    PermissionMode,
    SurfaceApprovalCommand,
    SurfaceBeginTurnCommand,
    SurfaceBeginTurnResponse,
    SurfaceClientRef,
    SurfaceCorrectionCommand,
    SurfaceEventBatch,
    SurfaceOpenSessionCommand,
    SurfaceSessionSnapshot,
    SurfaceSetPermissionModeCommand,
    SurfaceStreamBatch,
    SurfaceStreamFrame,
    SurfaceStreamSubscription,
    SurfaceTurnCommand,
    SurfaceTurnResponse,
    TaskEvent,
    canonical_json,
)

from apps.runtime_daemon.descriptor import RuntimeDescriptor


class SurfaceClientError(RuntimeError):
    """Closed client-side error base (never carries secrets)."""


class SurfaceProtocolMismatch(SurfaceClientError):
    """The server responded with an unsupported protocol version."""


class SurfaceClientAuthenticationError(SurfaceClientError):
    """The local runtime rejected the descriptor bearer token."""


class SurfaceClientConnectionError(SurfaceClientError):
    """The local runtime could not be reached."""


class SurfaceHttpError(SurfaceClientError):
    """The local runtime returned a non-success status."""

    def __init__(self, status_code: int, message: str) -> None:
        super().__init__(message)
        self.status_code = status_code


class SurfaceStreamStaleError(SurfaceClientError):
    """The transient stream cursor is stale: dead daemon generation or a
    stream the current runtime does not host. Resubscribe and re-sync from
    durable session state; transient frames are never replayable."""


class SurfaceClient:
    """Loopback client over the authenticated Surface HTTP/SSE protocol."""

    def __init__(self, descriptor: RuntimeDescriptor) -> None:
        self._base_url = descriptor.base_url
        self._token = descriptor.bearer_token
        self._hostname = socket.gethostname() or "local"
        self._device_id = f"device:{uuid4().hex[:8]}"
        self._sequences: dict[str, int] = {}

    def _client_ref(self) -> SurfaceClientRef:
        return SurfaceClientRef(
            client_id=f"cli:{self._hostname}",
            client_type="CLI",
            principal_id="user:local",
            tenant_id="tenant:local",
            workspace_id="workspace:local",
            device_id=self._device_id,
        )

    def _now(self) -> datetime:
        return datetime.now(timezone.utc)

    def _track(self, session_id: str, snapshot: SurfaceSessionSnapshot) -> None:
        self._sequences[session_id] = snapshot.event_sequence

    def _sequence(self, session_id: str) -> int:
        return self._sequences.get(session_id, 0)

    def _request(self, method: str, path: str, payload: ContractModel | None) -> dict:
        data = None if payload is None else canonical_json(payload).encode("utf-8")
        request = urllib.request.Request(
            self._base_url + path,
            data=data,
            method=method,
            headers={
                "Authorization": f"Bearer {self._token}",
                "Content-Type": "application/json",
                "X-Agent-OS-Protocol": SURFACE_PROTOCOL_VERSION,
            },
        )
        return self._decode_json(request)

    def _decode_json(self, request: urllib.request.Request) -> dict:
        try:
            with urllib.request.urlopen(request) as response:
                raw = response.read()
        except urllib.error.HTTPError as exc:
            status = exc.code
            try:
                body = json.loads(exc.read())
            except (ValueError, UnicodeDecodeError):
                body = {}
            message = body.get("message") or body.get("error") or f"HTTP {status}"
            if status == 401:
                raise SurfaceClientAuthenticationError(message) from exc
            raise SurfaceHttpError(status, message) from exc
        except urllib.error.URLError as exc:
            raise SurfaceClientConnectionError(
                f"cannot reach the local runtime: {exc.reason}"
            ) from exc
        try:
            value = json.loads(raw)
        except (ValueError, UnicodeDecodeError) as exc:
            raise SurfaceProtocolMismatch(
                "local runtime returned a non-JSON response"
            ) from exc
        if not isinstance(value, dict):
            raise SurfaceProtocolMismatch(
                "local runtime returned a non-object response"
            )
        return value

    def _check_protocol(self, value: object) -> None:
        if (
            not isinstance(value, dict)
            or value.get("protocol_version") != SURFACE_PROTOCOL_VERSION
        ):
            raise SurfaceProtocolMismatch(
                f"local runtime protocol is not {SURFACE_PROTOCOL_VERSION}"
            )

    def open_session(
        self, statement: str, *, idempotency_key: str | None = None
    ) -> SurfaceSessionSnapshot:
        if not statement.strip():
            raise ValueError("statement must be non-empty")
        command = SurfaceOpenSessionCommand(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            client=self._client_ref(),
            statement=statement,
            idempotency_key=idempotency_key or f"cli-open:{uuid4().hex}",
            requested_at=self._now(),
        )
        response = self._request("POST", "/v1/surface/sessions", command)
        snapshot_value = response.get("snapshot")
        if not isinstance(snapshot_value, dict):
            raise SurfaceProtocolMismatch("local runtime returned no session snapshot")
        self._check_protocol(snapshot_value)
        snapshot = SurfaceSessionSnapshot.model_validate(snapshot_value)
        self._track(snapshot.session.session_id, snapshot)
        return snapshot

    def get_session(self, session_id: str) -> SurfaceSessionSnapshot:
        if not session_id.strip():
            raise ValueError("session_id must be non-empty")
        response = self._request("GET", f"/v1/surface/sessions/{session_id}", None)
        self._check_protocol(response)
        snapshot = SurfaceSessionSnapshot.model_validate(response)
        self._track(snapshot.session.session_id, snapshot)
        return snapshot

    def run_turn(
        self,
        session_id: str,
        text: str,
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceTurnResponse:
        if not text.strip():
            raise ValueError("turn text must be non-empty")
        command = SurfaceTurnCommand(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            client=self._client_ref(),
            session_id=session_id,
            text=text,
            expected_event_sequence=(
                self._sequence(session_id)
                if expected_event_sequence is None
                else expected_event_sequence
            ),
            idempotency_key=idempotency_key or f"cli-turn:{uuid4().hex}",
            requested_at=self._now(),
        )
        response = self._request(
            "POST", f"/v1/surface/sessions/{session_id}/turns", command
        )
        return self._decode_turn(response)

    def decide_approval(
        self,
        session_id: str,
        action_digest: str,
        disposition: Literal[ApprovalDisposition.APPROVE, ApprovalDisposition.REJECT],
        reason: str,
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceTurnResponse:
        if not action_digest.strip() or not reason.strip():
            raise ValueError("action digest and reason must be non-empty")
        command = SurfaceApprovalCommand(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            client=self._client_ref(),
            session_id=session_id,
            action_digest=action_digest,
            disposition=disposition,
            reason=reason,
            expected_event_sequence=(
                self._sequence(session_id)
                if expected_event_sequence is None
                else expected_event_sequence
            ),
            idempotency_key=idempotency_key or f"cli-approve:{uuid4().hex}",
            requested_at=self._now(),
        )
        response = self._request(
            "POST", f"/v1/surface/sessions/{session_id}/approvals", command
        )
        return self._decode_turn(response)

    def pause(
        self,
        session_id: str,
        reason: str = "paused by user",
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceSessionSnapshot:
        return self._control(
            session_id,
            "pause",
            reason,
            expected_event_sequence=expected_event_sequence,
            idempotency_key=idempotency_key,
        )

    def resume(
        self,
        session_id: str,
        reason: str = "resumed by user",
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceSessionSnapshot:
        return self._control(
            session_id,
            "resume",
            reason,
            expected_event_sequence=expected_event_sequence,
            idempotency_key=idempotency_key,
        )

    def correct(
        self,
        session_id: str,
        reason: str,
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceSessionSnapshot:
        if not reason.strip():
            raise ValueError("correction reason must be non-empty")
        return self._control(
            session_id,
            "correction",
            reason,
            expected_event_sequence=expected_event_sequence,
            idempotency_key=idempotency_key,
        )

    def set_permission_mode(
        self,
        session_id: str,
        mode: PermissionMode,
        *,
        expected_event_sequence: int | None = None,
        idempotency_key: str | None = None,
    ) -> SurfaceSessionSnapshot:
        """E2 operator-issued mode change (the model can never set a mode)."""
        command = SurfaceSetPermissionModeCommand(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            client=self._client_ref(),
            session_id=session_id,
            mode=mode,
            expected_event_sequence=(
                self._sequence(session_id)
                if expected_event_sequence is None
                else expected_event_sequence
            ),
            idempotency_key=idempotency_key or f"cli-mode:{uuid4().hex}",
            requested_at=self._now(),
        )
        response = self._request(
            "POST", f"/v1/surface/sessions/{session_id}/mode", command
        )
        snapshot_value = response.get("snapshot")
        if not isinstance(snapshot_value, dict):
            raise SurfaceProtocolMismatch("local runtime returned no session snapshot")
        self._check_protocol(snapshot_value)
        snapshot = SurfaceSessionSnapshot.model_validate(snapshot_value)
        self._track(session_id, snapshot)
        return snapshot

    def _control(
        self,
        session_id: str,
        action: str,
        reason: str,
        *,
        expected_event_sequence: int | None,
        idempotency_key: str | None,
    ) -> SurfaceSessionSnapshot:
        command = SurfaceCorrectionCommand(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            client=self._client_ref(),
            session_id=session_id,
            reason=reason,
            expected_event_sequence=(
                self._sequence(session_id)
                if expected_event_sequence is None
                else expected_event_sequence
            ),
            idempotency_key=idempotency_key or f"cli-{action}:{uuid4().hex}",
            requested_at=self._now(),
        )
        response = self._request(
            "POST", f"/v1/surface/sessions/{session_id}/{action}", command
        )
        snapshot_value = response.get("snapshot")
        if not isinstance(snapshot_value, dict):
            raise SurfaceProtocolMismatch("local runtime returned no session snapshot")
        self._check_protocol(snapshot_value)
        snapshot = SurfaceSessionSnapshot.model_validate(snapshot_value)
        self._track(session_id, snapshot)
        return snapshot

    def _decode_turn(self, response: dict) -> SurfaceTurnResponse:
        turn_value = response.get("turn")
        if not isinstance(turn_value, dict):
            raise SurfaceProtocolMismatch("local runtime returned no turn response")
        self._check_protocol(turn_value)
        snapshot_value = turn_value.get("snapshot")
        if isinstance(snapshot_value, dict):
            self._check_protocol(snapshot_value)
        turn = SurfaceTurnResponse.model_validate(turn_value)
        if turn.snapshot.session.session_id:
            self._track(turn.snapshot.session.session_id, turn.snapshot)
        return turn

    def events(
        self,
        task_id: str,
        *,
        after_sequence: int = 0,
        wait_ms: int = 0,
    ) -> SurfaceEventBatch:
        if not task_id.strip():
            raise ValueError("task_id must be non-empty")
        path = (
            f"/v1/surface/tasks/{task_id}/events"
            f"?after={after_sequence}&wait_ms={wait_ms}"
        )
        body = self._request_raw("GET", path)
        return self._decode_sse(task_id, after_sequence, body)

    def _request_raw(self, method: str, path: str) -> bytes:
        request = urllib.request.Request(
            self._base_url + path,
            method=method,
            headers={
                "Authorization": f"Bearer {self._token}",
                "X-Agent-OS-Protocol": SURFACE_PROTOCOL_VERSION,
            },
        )
        try:
            with urllib.request.urlopen(request) as response:
                return response.read()
        except urllib.error.HTTPError as exc:
            status = exc.code
            try:
                body = json.loads(exc.read())
            except (ValueError, UnicodeDecodeError):
                body = {}
            message = body.get("message") or body.get("error") or f"HTTP {status}"
            if status == 401:
                raise SurfaceClientAuthenticationError(message) from exc
            if status == 410:
                raise SurfaceStreamStaleError(message) from exc
            raise SurfaceHttpError(status, message) from exc
        except urllib.error.URLError as exc:
            raise SurfaceClientConnectionError(
                f"cannot reach the local runtime: {exc.reason}"
            ) from exc

    def subscribe_stream(self, session_id: str) -> SurfaceStreamSubscription:
        """Subscription-first: mint a transient stream under the current
        daemon generation before executing any turn (E1, frozen order)."""
        if not session_id.strip():
            raise ValueError("session_id must be non-empty")
        payload = self._request(
            "POST", f"/v1/surface/sessions/{session_id}/streams", None
        )
        subscription = payload.get("subscription")
        if not isinstance(subscription, dict):
            raise SurfaceProtocolMismatch(
                "local runtime returned no stream subscription"
            )
        self._check_protocol(subscription)
        return SurfaceStreamSubscription.model_validate(subscription)

    def begin_turn(self, command: SurfaceBeginTurnCommand) -> SurfaceBeginTurnResponse:
        """E1 reserve/begin-turn: bind the pre-subscribed stream and return
        the authoritative {turn_id, stream_id} without blocking on the
        provider stream."""
        payload = self._request(
            "POST",
            f"/v1/surface/sessions/{command.session_id}/begin-turn",
            command,
        )
        begin_turn = payload.get("begin_turn")
        if not isinstance(begin_turn, dict):
            raise SurfaceProtocolMismatch(
                "local runtime returned no begin-turn response"
            )
        self._check_protocol(begin_turn)
        return SurfaceBeginTurnResponse.model_validate(begin_turn)

    def stream_frames(
        self,
        session_id: str,
        *,
        stream_id: str,
        runtime_boot_id: str,
        after_sequence: int = 0,
        wait_ms: int = 0,
        last_event_id: int | None = None,
    ) -> SurfaceStreamBatch:
        """Bounded incremental read of transient display frames (E1).

        410 maps to typed `SurfaceStreamStaleError`: the generation is dead or
        the stream is unknown; resubscribe, never replay.
        """
        if not session_id.strip():
            raise ValueError("session_id must be non-empty")
        if not stream_id.strip():
            raise ValueError("stream_id must be non-empty")
        if not runtime_boot_id.strip():
            raise ValueError("runtime_boot_id must be non-empty")
        cursor = last_event_id if last_event_id is not None else after_sequence
        path = (
            f"/v1/surface/sessions/{session_id}/stream"
            f"?stream_id={stream_id}&after={cursor}&wait_ms={wait_ms}"
        )
        body = self._request_raw("GET", path)
        return self._decode_frame_sse(session_id, cursor, body)

    @staticmethod
    def _decode_frame_sse(
        session_id: str, after_sequence: int, body: bytes
    ) -> SurfaceStreamBatch:
        frames: list[SurfaceStreamFrame] = []
        next_sequence = after_sequence
        current_event: str | None = None
        data_lines: list[str] = []
        for line in body.decode("utf-8", errors="strict").splitlines():
            if line.startswith("event: "):
                current_event = line[7:]
            elif line.startswith("data: "):
                data_lines.append(line[6:])
            elif line == "":
                if current_event == "cursor":
                    cursor_payload = json.loads("\n".join(data_lines))
                    next_sequence = cursor_payload["next_sequence"]
                elif current_event is not None and data_lines:
                    frames.append(
                        SurfaceStreamFrame.model_validate(
                            json.loads("\n".join(data_lines))
                        )
                    )
                current_event = None
                data_lines = []
        return SurfaceStreamBatch(
            session_id=session_id,
            after_sequence=after_sequence,
            next_sequence=next_sequence,
            frames=tuple(frames),
        )

    @staticmethod
    def _decode_sse(
        task_id: str, after_sequence: int, body: bytes
    ) -> SurfaceEventBatch:
        events: list[TaskEvent] = []
        current_id: int | None = None
        data_lines: list[str] = []
        in_cursor = False
        next_sequence = after_sequence
        for line in body.decode("utf-8", errors="strict").splitlines():
            if line.startswith("id: "):
                current_id = int(line[4:])
            elif line.startswith("event: cursor"):
                in_cursor = True
                data_lines = []
            elif line.startswith("data: "):
                data_lines.append(line[6:])
            elif line == "":
                if in_cursor:
                    cursor_payload = json.loads("\n".join(data_lines))
                    next_sequence = cursor_payload["next_sequence"]
                    in_cursor = False
                elif current_id is not None and data_lines:
                    payload = json.loads("\n".join(data_lines))
                    events.append(TaskEvent.model_validate(payload))
                current_id = None
                data_lines = []
        return SurfaceEventBatch(
            protocol_version=SURFACE_PROTOCOL_VERSION,
            task_id=task_id,
            after_sequence=after_sequence,
            next_sequence=next_sequence,
            events=tuple(events),
        )

File: apps/macos/renderer/src/surface_client.ts
// Typed Surface HTTP/SSE client for the Wave 2a renderer.
//
// Mirrors apps/cli/surface_client.py semantics: every success is parsed
// through the matching closed contract, the protocol version is checked on
// every response, HTTP errors map to closed typed errors, and the per-session
// event sequence is tracked for commands. The bearer token is supplied by the
// Rust layer and never stored in webview-local storage.

export const SURFACE_PROTOCOL_VERSION = "1.1";

export class SurfaceProtocolMismatch extends Error {}
export class SurfaceClientAuthenticationError extends Error {}
export class SurfaceClientConnectionError extends Error {}
export class SurfaceHttpError extends Error {
  statusCode: number;
  constructor(statusCode: number, message: string) {
    super(message);
    this.statusCode = statusCode;
  }
}

export interface FetchLike {
  (input: string, init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: string;
  }): Promise<{ ok: boolean; status: number; text(): Promise<string> }>;
}

export interface SessionRef {
  session_id: string;
  task_id: string;
  run_id: string;
  tenant_id: string;
  workspace_id: string;
}

export interface PendingApproval {
  action_digest: string;
  capability_id: string;
  proposal_id: string;
  preview: string;
  requested_at: string;
}

export interface SessionSnapshot {
  protocol_version: string;
  session: SessionRef;
  envelope_id: string;
  expected_outcome_id: string;
  status: string;
  event_sequence: number;
  message_count: number;
  pending_approval: PendingApproval | null;
  updated_at: string;
}

export interface TurnResponse {
  protocol_version: string;
  snapshot: SessionSnapshot;
  turn_id: string | null;
  text: string;
  steps: unknown[];
  stop_reason: string;
  total_tokens: number;
}

export interface SurfaceEvent {
  sequence: number;
  event_type: string;
  payload_json: string;
}

export interface EventBatch {
  protocol_version: string;
  task_id: string;
  after_sequence: number;
  next_sequence: number;
  events: SurfaceEvent[];
}

/**
 * Mirror of SurfaceEventBatch._validate_event_ownership_and_sequence
 * (packages/contracts/src/agent_os_contracts/surface.py): the resume cursor in
 * a durable event batch must agree with the events that batch carries. Callers
 * (agent_thread) resume from `next_sequence`, so accepting a self-contradictory
 * batch silently skips events or re-requests an already-applied window.
 */
function checkEventSequence(
  afterSequence: number,
  nextSequence: number,
  events: readonly SurfaceEvent[],
): void {
  let previousSequence = afterSequence;
  for (const event of events) {
    if (event.sequence <= previousSequence) {
      throw new SurfaceProtocolMismatch(
        "surface event sequences must strictly increase above after_sequence",
      );
    }
    previousSequence = event.sequence;
  }
  if (nextSequence !== previousSequence) {
    throw new SurfaceProtocolMismatch(
      "surface next_sequence must equal the last event sequence or after_sequence",
    );
  }
}

export interface ConflictProjection {
  protocol_version: string;
  action_id: string;
  lease_id: string;
  disposition: "REPLAN" | "CONFLICT" | "CANCEL";
  reason: string;
  write_scope_uris: string[];
  relevant_event_ids: string[];
  suggested_action: "REPLAN" | "REVIEW_DIFF" | "NONE";
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function checkProtocol(value: Record<string, unknown>): void {
  if (value.protocol_version !== SURFACE_PROTOCOL_VERSION) {
    throw new SurfaceProtocolMismatch(
      `local runtime protocol is not ${SURFACE_PROTOCOL_VERSION}`,
    );
  }
}

function checkSnapshot(value: Record<string, unknown>): SessionSnapshot {
  checkProtocol(value);
  if (!isObject(value.session) || typeof value.session.session_id !== "string") {
    throw new SurfaceProtocolMismatch("local runtime returned no session snapshot");
  }
  const session = value.session as unknown as SessionRef;
  const pending = isObject(value.pending_approval)
    ? (value.pending_approval as unknown as PendingApproval)
    : null;
  return {
    protocol_version: value.protocol_version as string,
    session,
    envelope_id: value.envelope_id as string,
    expected_outcome_id: value.expected_outcome_id as string,
    status: value.status as string,
    event_sequence: value.event_sequence as number,
    message_count: value.message_count as number,
    pending_approval: pending,
    updated_at: value.updated_at as string,
  };
}

function checkTurn(value: Record<string, unknown>): TurnResponse {
  checkProtocol(value);
  if (!isObject(value.snapshot)) {
    throw new SurfaceProtocolMismatch("local runtime returned no turn response");
  }
  return {
    protocol_version: value.protocol_version as string,
    snapshot: checkSnapshot(value.snapshot as Record<string, unknown>),
    turn_id: value.turn_id === null ? null : (value.turn_id as string),
    text: value.text as string,
    steps: Array.isArray(value.steps) ? (value.steps as unknown[]) : [],
    stop_reason: value.stop_reason as string,
    total_tokens: value.total_tokens as number,
  };
}

export class SurfaceClient {
  private baseUrl: string;
  private token: string;
  private fetchImpl: FetchLike;
  private sequences: Map<string, number> = new Map();

  constructor(
    baseUrl: string,
    token: string,
    fetchImpl: FetchLike = (input, init) => fetch(input, init as RequestInit),
  ) {
    this.baseUrl = baseUrl;
    this.token = token;
    this.fetchImpl = fetchImpl;
  }

  baseUrlFor(): string {
    return this.baseUrl;
  }

  tokenFor(): string {
    return this.token;
  }

  private track(sessionId: string, snapshot: SessionSnapshot): void {
    this.sequences.set(sessionId, snapshot.event_sequence);
  }

  private sequence(sessionId: string): number {
    return this.sequences.get(sessionId) ?? 0;
  }

  private async requestJson(
    method: string,
    path: string,
    payload: Record<string, unknown> | null,
  ): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.token}`,
      "Content-Type": "application/json",
      "X-Agent-OS-Protocol": SURFACE_PROTOCOL_VERSION,
    };
    let response;
    try {
      response = await this.fetchImpl(this.baseUrl + path, {
        method,
        headers,
        body: payload === null ? undefined : JSON.stringify(payload),
      });
    } catch (error) {
      throw new SurfaceClientConnectionError(
        `cannot reach the local runtime: ${String(error)}`,
      );
    }
    const raw = await response.text();
    if (!response.ok) {
      let message = `HTTP ${response.status}`;
      try {
        const body = JSON.parse(raw) as Record<string, unknown>;
        if (typeof body.message === "string") message = body.message;
        else if (typeof body.error === "string") message = body.error;
      } catch {
        // non-JSON error body; keep the default message
      }
      if (response.status === 401) {
        throw new SurfaceClientAuthenticationError(message);
      }
      throw new SurfaceHttpError(response.status, message);
    }
    let value: unknown;
    try {
      value = JSON.parse(raw);
    } catch {
      throw new SurfaceProtocolMismatch(
        "local runtime returned a non-JSON response",
      );
    }
    if (!isObject(value)) {
      throw new SurfaceProtocolMismatch(
        "local runtime returned a non-object response",
      );
    }
    return value;
  }

  private clientRef(): Record<string, string> {
    return {
      client_id: "renderer",
      client_type: "DESKTOP",
      principal_id: "user:local",
      tenant_id: "tenant:local",
      workspace_id: "workspace:local",
      device_id: "device:renderer",
    };
  }

  async openSession(statement: string): Promise<SessionSnapshot> {
    const response = await this.requestJson("POST", "/v1/surface/sessions", {
      protocol_version: SURFACE_PROTOCOL_VERSION,
      client: this.clientRef(),
      statement,
      idempotency_key: `renderer-open-${crypto.randomUUID()}`,
      requested_at: new Date().toISOString(),
    });
    const snapshot = isObject(response.snapshot)
      ? checkSnapshot(response.snapshot as Record<string, unknown>)
      : (() => {
          throw new SurfaceProtocolMismatch(
            "local runtime returned no session snapshot",
          );
        })();
    this.track(snapshot.session.session_id, snapshot);
    return snapshot;
  }

  async getSession(sessionId: string): Promise<SessionSnapshot> {
    const response = await this.requestJson(
      "GET",
      `/v1/surface/sessions/${encodeURIComponent(sessionId)}`,
      null,
    );
    const snapshot = checkSnapshot(response);
    this.track(snapshot.session.session_id, snapshot);
    return snapshot;
  }

  async runTurn(sessionId: string, text: string): Promise<TurnResponse> {
    const response = await this.requestJson(
      "POST",
      `/v1/surface/sessions/${encodeURIComponent(sessionId)}/turns`,
      {
        protocol_version: SURFACE_PROTOCOL_VERSION,
        client: this.clientRef(),
        session_id: sessionId,
        text,
        expected_event_sequence: this.sequence(sessionId),
        idempotency_key: `renderer-turn-${crypto.randomUUID()}`,
        requested_at: new Date().toISOString(),
      },
    );
    const turn = isObject(response.turn)
      ? checkTurn(response.turn as Record<string, unknown>)
      : (() => {
          throw new SurfaceProtocolMismatch(
            "local runtime returned no turn response",
          );
        })();
    this.track(turn.snapshot.session.session_id, turn.snapshot);
    return turn;
  }

  async decideApproval(
    sessionId: string,
    actionDigest: string,
    disposition: "APPROVE" | "REJECT",
    reason: string,
  ): Promise<TurnResponse> {
    const response = await this.requestJson(
      "POST",
      `/v1/surface/sessions/${encodeURIComponent(sessionId)}/approvals`,
      {
        protocol_version: SURFACE_PROTOCOL_VERSION,
        client: this.clientRef(),
        session_id: sessionId,
        action_digest: actionDigest,
        disposition,
        reason,
        expected_event_sequence: this.sequence(sessionId),
        idempotency_key: `renderer-approve-${crypto.randomUUID()}`,
        requested_at: new Date().toISOString(),
      },
    );
    const turn = isObject(response.turn)
      ? checkTurn(response.turn as Record<string, unknown>)
      : (() => {
          throw new SurfaceProtocolMismatch(
            "local runtime returned no turn response",
          );
        })();
    this.track(turn.snapshot.session.session_id, turn.snapshot);
    return turn;
  }

  async pause(sessionId: string, reason = "paused by user"): Promise<SessionSnapshot> {
    return this.control(sessionId, "pause", reason);
  }

  async resume(sessionId: string, reason = "resumed by user"): Promise<SessionSnapshot> {
    return this.control(sessionId, "resume", reason);
  }

  async correct(sessionId: string, reason: string): Promise<SessionSnapshot> {
    return this.control(sessionId, "correction", reason);
  }

  private async control(
    sessionId: string,
    action: string,
    reason: string,
  ): Promise<SessionSnapshot> {
    const response = await this.requestJson(
      "POST",
      `/v1/surface/sessions/${encodeURIComponent(sessionId)}/${action}`,
      {
        protocol_version: SURFACE_PROTOCOL_VERSION,
        client: this.clientRef(),
        session_id: sessionId,
        reason,
        expected_event_sequence: this.sequence(sessionId),
        idempotency_key: `renderer-${action}-${crypto.randomUUID()}`,
        requested_at: new Date().toISOString(),
      },
    );
    const snapshot = isObject(response.snapshot)
      ? checkSnapshot(response.snapshot as Record<string, unknown>)
      : (() => {
          throw new SurfaceProtocolMismatch(
            "local runtime returned no session snapshot",
          );
        })();
    this.track(sessionId, snapshot);
    return snapshot;
  }

  async events(
    taskId: string,
    afterSequence = 0,
    waitMs = 0,
  ): Promise<EventBatch> {
    let response;
    try {
      response = await this.fetchImpl(
        `${this.baseUrl}/v1/surface/tasks/${encodeURIComponent(taskId)}/events?after=${afterSequence}&wait_ms=${waitMs}`,
        {
          method: "GET",
          headers: {
            Authorization: `Bearer ${this.token}`,
            "X-Agent-OS-Protocol": SURFACE_PROTOCOL_VERSION,
          },
        },
      );
    } catch (error) {
      throw new SurfaceClientConnectionError(
        `cannot reach the local runtime: ${String(error)}`,
      );
    }
    const body = await response.text();
    if (!response.ok) {
      if (response.status === 401) {
        throw new SurfaceClientAuthenticationError("local runtime authentication failed");
      }
      throw new SurfaceHttpError(response.status, `HTTP ${response.status}`);
    }
    return parseSse(taskId, afterSequence, body);
  }

  async conflict(sessionId: string): Promise<ConflictProjection | null> {
    let response;
    try {
      response = await this.fetchImpl(
        `${this.baseUrl}/v1/surface/sessions/${encodeURIComponent(sessionId)}/conflict`,
        {
          method: "GET",
          headers: {
            Authorization: `Bearer ${this.token}`,
            "X-Agent-OS-Protocol": SURFACE_PROTOCOL_VERSION,
          },
        },
      );
    } catch (error) {
      throw new SurfaceClientConnectionError(
        `cannot reach the local runtime: ${String(error)}`,
      );
    }
    const raw = await response.text();
    if (response.status === 404) {
      return null;
    }
    if (!response.ok) {
      throw new SurfaceHttpError(response.status, `HTTP ${response.status}`);
    }
    let value: unknown;
    try {
      value = JSON.parse(raw);
    } catch {
      throw new SurfaceProtocolMismatch(
        "local runtime returned a non-JSON response",
      );
    }
    if (!isObject(value) || !isObject(value.conflict)) {
      throw new SurfaceProtocolMismatch(
        "local runtime returned no conflict projection",
      );
    }
    return checkConflict(value.conflict as Record<string, unknown>);
  }
}

function checkConflict(value: Record<string, unknown>): ConflictProjection {
  checkProtocol(value);
  const disposition = value.disposition;
  const suggested = value.suggested_action;
  if (
    disposition !== "REPLAN" &&
    disposition !== "CONFLICT" &&
    disposition !== "CANCEL"
  ) {
    throw new SurfaceProtocolMismatch("conflict projection has invalid disposition");
  }
  if (
    suggested !== "REPLAN" &&
    suggested !== "REVIEW_DIFF" &&
    suggested !== "NONE"
  ) {
    throw new SurfaceProtocolMismatch("conflict projection has invalid suggested_action");
  }
  return {
    protocol_version: value.protocol_version as string,
    action_id: value.action_id as string,
    lease_id: value.lease_id as string,
    disposition,
    reason: value.reason as string,
    write_scope_uris: Array.isArray(value.write_scope_uris)
      ? (value.write_scope_uris as string[])
      : [],
    relevant_event_ids: Array.isArray(value.relevant_event_ids)
      ? (value.relevant_event_ids as string[])
      : [],
    suggested_action: suggested,
  };
}

export function parseSse(
  taskId: string,
  afterSequence: number,
  body: string,
): EventBatch {
  const events: SurfaceEvent[] = [];
  let currentId: number | null = null;
  const dataLines: string[] = [];
  let inCursor = false;
  let nextSequence = afterSequence;
  const flush = (): void => {
    if (inCursor) {
      const cursor = JSON.parse(dataLines.join("\n")) as {
        next_sequence: number;
      };
      nextSequence = cursor.next_sequence;
      inCursor = false;
    } else if (currentId !== null && dataLines.length > 0) {
      const payload = JSON.parse(dataLines.join("\n")) as {
        event_type: string;
        payload_json: string;
      };
      events.push({
        sequence: currentId,
        event_type: payload.event_type,
        payload_json: payload.payload_json,
      });
    }
    currentId = null;
    dataLines.length = 0;
  };
  for (const line of body.split(/\r?\n/)) {
    if (line.startsWith("id: ")) {
      currentId = Number.parseInt(line.slice(4), 10);
    } else if (line.startsWith("event: cursor")) {
      inCursor = true;
      dataLines.length = 0;
    } else if (line.startsWith("data: ")) {
      dataLines.push(line.slice(6));
    } else if (line === "") {
      flush();
    }
  }
  // EOF flush, mirroring apps/cli/surface_client.py `_decode_sse` and
  // apps/cli-ts/src/sse.ts: a frame is normally terminated by a blank line, but
  // a body that ends mid-frame must be decoded all the same. Dropping the final
  // cursor left next_sequence at after_sequence; a truncated payload still
  // fails closed because JSON.parse rejects it.
  flush();
  checkEventSequence(afterSequence, nextSequence, events);
  return {
    protocol_version: SURFACE_PROTOCOL_VERSION,
    task_id: taskId,
    after_sequence: afterSequence,
    next_sequence: nextSequence,
    events,
  };
}
